cpp_generator.py

```diff
@@ -0,0 +1,347 @@
+        with class_code_context as public_scope:
+            if name.endswith("Collection"):
+                public_scope.write("%NESTED_CLASSES%\n")
+            public_scope.write(f"{class_name}({ctor_params})")
+            if len(ctor_initializer_list) > 0:
+                with public_scope as ctor_initializer_list_scope:
+                    ctor_initializer_list_scope.write(f":\n{ctor_initializer_list}\n")
+                public_scope.write("{\n}\n\n")
+
+            if len(method_list) > 0:
+                public_scope.write("\n\n".join(method_list))
+                public_scope.write("\n")
+
+            member_list = ";\n".join(member_list)
+            public_scope.write(f"{member_list}" + ";\n")
+
+        class_code_context.write("};\n")
+        return class_code_context.get_content()
+
+    def __gen_collection_types(self, node: VSSNode, path: str) -> str:
+        collection_types = []
+        for child in node.children:
+            if child.type == VSSType.BRANCH:
+                self.includes.add(f"{path}/{child.name}/{child.name}")
+
+                if child.instances:
+                    instances = [(f"{child.name}Collection", None)] + self.__gen_instances(child)
+                    generated_classes = []
+
+                    # create all nested classes for this sub-tree
+                    for i in range(len(instances) - 1):
+                        nested_class = self.__gen_nested_class(child, instances, i)
+                        generated_classes.append(nested_class)
+                        
+                    collection_types.append(generated_classes[0].replace("%NESTED_CLASSES%", "\n\n".join(generated_classes[1:])))
+        
+        return "\n\n".join(collection_types)
+
+    def __gen_model(self, node: VSSNode, path: str, is_root=False):
+        # must be done before generating the imports, since it is adding imports to the list
+        collection_types = self.__gen_collection_types(node, path)
+
+        self.__gen_header(node)
+        self.__gen_imports(node)
+        self.ctx_header.write(self.__generate_opening_namespace_text(self.root_namespace, node))
+        self.ctx_header.write("using ParentClass = Model;\n\n")
+        self.__gen_model_docstring(node)
+        self.ctx_header.write(f"class {node.name} : public ParentClass {{\n")
+        self.ctx_header.write("public:\n")
+
+        with self.ctx_header as header_public:
+            header_public.write(collection_types)
+            header_public.write("\n")
+
+            if is_root:
+                header_public.write(f"{node.name}() :\n")
+                header_public.indent()
+                header_public.write("ParentClass(\"Vehicle\")")
+            else:
+                header_public.write(f"{node.name}(const std::string& name, ParentClass* parent) :\n")
+                header_public.indent() 
+                header_public.write("ParentClass(name, parent)")
+            
+            header_public.write("%MEMBER%\n")
```

Comment:
Why do we need the `%MEMBER%` placeholder here? Whenever possible I suggest we write the content directly and not mix 2 styles. Otherwise we might want to switch to an entirely template based approach.

Reply:
agreed, but to get all the members, we need to iterate through all the children. This is done afterwards and I wanted to avoid the same loop again.

Reply:
it is the same as for the nested classes